inlong-sort/sort-flink/sort-flink-v1.13/sort-connectors/iceberg/src/main/java/org/apache/inlong/sort/iceberg/sink/multiple/SchemaChangeUtils.java

```diff
@@ -51,49 +53,81 @@ public class SchemaChangeUtils {
     static List<TableChange> diffSchema(Schema oldSchema, Schema newSchema) {
         List<String> oldFields = oldSchema.columns().stream().map(NestedField::name).collect(Collectors.toList());
         List<String> newFields = newSchema.columns().stream().map(NestedField::name).collect(Collectors.toList());
-        int oi = 0;
-        int ni = 0;
+        Set<String> oldFieldSet = new HashSet<>(oldFields);
+        Set<String> newFieldSet = new HashSet<>(newFields);
+
+        Set<String> intersectColSet = Sets.intersection(oldFieldSet, newFieldSet);
+        Set<String> colsToDelete = Sets.difference(oldFieldSet, newFieldSet);
+        Set<String> colsToAdd = Sets.difference(newFieldSet, oldFieldSet);
+
         List<TableChange> tableChanges = new ArrayList<>();
-        while (ni < newFields.size()) {
-            if (oi < oldFields.size() && oldFields.get(oi).equals(newFields.get(ni))) {
-                oi++;
-                ni++;
-            } else {
-                NestedField newField = newSchema.findField(newFields.get(ni));
+
+        // step0: Unknown change
+        if (!colsToDelete.isEmpty() && !colsToAdd.isEmpty()) {
```

Comment:
because it can not distinguished (add + delete) or modify, example first [a, b, c] -> delete c [a, b] -> add d [a, b, d], it not suppoted to judge. I will support later use information extracted from DDL.

Reply:
> example first [a, b, c] -> delete c [a, b] -> add d [a, b, d]

If two messages are received one after another: (a1,b1,c1)->(a1,b1,d1), through the diff schema, you can get an add(d) and delete(c), so after executing add and delete , becomes [a, b, d] ?

Reply:
of course not,for iceberg, each field has a unique id. Even if it is deleted, the field still exists, but the content of the field will be hidden when querying.if modify, only one modify step [a, b, c] -> [a, b, d],The field[d] id has not changed.if add and delete, the filed d has a new unique id. So first add and then delete cannot be confused with modify.

Reply:
I got it, thanks for clarifying